File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1: return 0 
        p1 = 0 ; p2 = 0
        cnt = 0
        product = 1
        for p2 in range(len(nums)):
            product = product * nums[p2]
            while product >= k:
                product /= nums[p1]
                p1 += 1
            cnt += p2 - p1 + 1
        return cnt
            
        

File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1)
            return 0;
        int p1 = 0 ; 
        int ans = 0 ;
        int total = 1 ; 
        for (int p2 = 0 ; p2 < nums.length;p2 ++){
            total = total * nums[p2];
            while (total >= k){
                total = total / nums[p1];
                p1 += 1 ;
            }
            ans += p2 - p1 + 1;
        }
        return ans ;
    }
}